apps/agreement/contracts/Agreement.sol

```diff
@@ -0,0 +1,1350 @@
+        return _existChallenge(_challengeId) && _actionId == _challenge.actionId;
+    }
+
+    /**
+    * @dev Tell whether a challenge exists
+    * @param _challengeId Identification number of the challenge being queried
+    * @return True if the requested challenge exists, false otherwise
+    */
+    function _existChallenge(uint256 _challengeId) internal view returns (bool) {
+        return _challengeId < challengesLength;
+    }
+
+    /**
+    * @dev Ensure a disputable entity is registered
+    * @param _disputableInfo Disputable info of the app being queried
+    */
+    function _ensureRegisteredDisputable(DisputableInfo storage _disputableInfo) internal view {
+        require(_disputableInfo.registered, ERROR_DISPUTABLE_APP_NOT_REGISTERED);
+    }
+
+    /**
+    * @dev Ensure a disputable entity is unregistered
+    * @param _disputableInfo Disputable info of the app being queried
+    */
+    function _ensureUnregisteredDisputable(DisputableInfo storage _disputableInfo) internal view {
+        require(!_disputableInfo.registered, ERROR_DISPUTABLE_APP_ALREADY_EXISTS);
+    }
+
+    /**
+    * @dev Tell the amount of leftover arbitration fees that the submitter must pay in order to to raise a dispute to the arbitrator
+    * @param _arbitrator Arbitrator to query
+    * @param _challengerFeeToken ERC20 token used for the arbitration fees paid by the challenger in advance
+    * @param _challengerFeeAmount Amount of arbitration fees paid by the challenger in advance
+    * @return Address where the arbitration fees must be transferred to
+    * @return ERC20 token to be used for the arbitration fees
+    * @return Amount of arbitration fees missing
+    * @return Total amount of arbitration fees required by the arbitrator to raise a dispute
+    */
+    function _getMissingArbitratorFees(IArbitrator _arbitrator, ERC20 _challengerFeeToken, uint256 _challengerFeeAmount) internal view
+        returns (address, ERC20, uint256, uint256)
+    {
+        (address disputeFeeRecipient, ERC20 feeToken, uint256 disputeFees) = _arbitrator.getDisputeFees();
+
+        uint256 missingFees;
+        if (_challengerFeeToken == feeToken) {
+            missingFees = _challengerFeeAmount >= disputeFees ? 0 : (disputeFees - _challengerFeeAmount);
+        } else {
+            missingFees = disputeFees;
+        }
+
+        return (disputeFeeRecipient, feeToken, missingFees, disputeFees);
+    }
+
+    /**
+    * @dev Helper to build an agreement dispute metadata as "agreements:[ACTION_ID]"
+    * @param _actionId Identification number of the action to create a dispute for
+    * @return dispute metadata for the requested action
+    */
+    function _buildDisputeMetadata(uint256 _actionId) internal view returns (bytes memory) {
+        // Header "agreement:"
+        bytes memory metadata = new bytes(11);
+        assembly { mstore(add(metadata, 32), 0x61677265656d656e74733A000000000000000000000000000000000000000000) }
```

Comment:
⚠️ I think this is being decoded as agreements at the moment (despite the comment).

But I also wonder if it would be better to use the Agreement's appID in string or namehash form instead? Maybe the namehash is best, as it's a short bytes32 value.

Reply:
Yup, let's do that